realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp

```diff
@@ -433,22 +435,83 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_SharedRealm_nativeGetTable(JNIEnv
 
 JNIEXPORT jlong JNICALL Java_io_realm_internal_SharedRealm_nativeCreateTable(JNIEnv* env, jclass,
                                                                              jlong shared_realm_ptr,
-                                                                             jstring table_name)
+                                                                             jstring j_table_name,
+                                                                             jboolean is_pk_table)
 {
     TR_ENTER_PTR(shared_realm_ptr)
 
-    std::string name_str;
+    std::string table_name;
     try {
-        JStringAccessor name(env, table_name); // throws
-        name_str = name;
+        table_name = JStringAccessor(env, j_table_name); // throws
         auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
         shared_realm->verify_in_write(); // throws
-        Table* table = LangBindHelper::add_table(shared_realm->read_group(), name); // throws
+        Table* table;
+        auto& group = shared_realm->read_group();
+        if (is_pk_table) {
+            // sync::create_table() will add an extra column for stable ID which is not allowed for pk table.
+            table = LangBindHelper::add_table(group, table_name); // throws
+        }
+        else {
+#if REALM_ENABLE_SYNC
+            // Sync doesn't throw when table exists.
+            if (group.has_table(table_name)) {
+                THROW_JAVA_EXCEPTION(
+                    env, JavaExceptionDef::IllegalArgument,
+                    format(c_table_name_exists_exception_msg, table_name.substr(TABLE_PREFIX.length())));
+            }
+            auto table_ref = sync::create_table(group, table_name); // throws
+            table = LangBindHelper::get_table(group, table_ref->get_index_in_group());
+#else
+            table = LangBindHelper::add_table(group, table_name); // throws
+#endif
+        }
+        return reinterpret_cast<jlong>(table);
+    }
+    catch (TableNameInUse& e) {
+        // We need to print the table name, so catch the exception here.
+        std::string class_name_str(table_name.substr(TABLE_PREFIX.length()));
```

Comment:
> Is it always OK to remove first a few characters?
yes, otherwise we made something really wrong.

> Is it sure that this method is not used for pk or metadata table?
This method is used for pk table as well, but we have check before, see 
https://github.com/realm/realm-java/pull/4693/files#diff-282e4aca0c4608335667e0c13e86e699L612

> I've found that there is a special method to create pk table.
> We don't have metadata table already?
The pk table is a bit messy, ideally it should always be handled in Object Store. And by moving move schema things there, we are quite close to the target.
meta table is better, we always calling `ObjectStore::set_schema_version()` which will create the meta table if it doesn't exist.